---

# 📒 EDA STRUCTURE – CREDIT RISK MODELING

---

## 🟦 TASK 0: Define the Problem (Business Understanding)

**Objective**

* Define what *default* means
* Define when prediction is made (application time)
* Define modeling goal (probability of default)

**Deliverables**

* Clear target variable definition
* Modeling type: Classification
* Business constraints (explainability, regulation)

---

## 🟦 TASK 1: Load Data & Initial Inspection

**What to do**

* Load dataset
* Inspect shape
* Inspect data types
* Preview sample rows

**Questions to answer**

* How many records?
* How many features?
* Any obvious data type issues?

**Deliverables**

* Dataset dimensions
* Data type summary
* Initial observations

---

## 🟦 TASK 2: Identify Target Variable

**What to do**

* Locate target column
* Check value counts
* Compute default rate

**Questions to answer**

* Is the target binary?
* Is the dataset imbalanced?

**Deliverables**

* Default vs non-default %
* Note on imbalance impact on metrics

---

## 🟦 TASK 3: Separate Feature Types

**What to do**

* Identify:

  * Numerical features
  * Categorical features
  * Date features
  * ID / non-predictive columns

**Questions to answer**

* Which columns should never go to the model?

**Deliverables**

* Feature categorization list

---

## 🟦 TASK 4: Missing Value Analysis

**What to do**

* Calculate missing % per column
* Visualize missingness
* Compare default rate for missing vs non-missing

**Questions to answer**

* Is missingness random?
* Does missingness signal higher risk?

**Deliverables**

* Columns with high missing %
* Missing value risk insights

---

## 🟦 TASK 5: Univariate Analysis – Numerical Features

**What to do**

* Plot distributions
* Check skewness
* Detect outliers
* Validate logical ranges

**Questions to answer**

* Are values realistic?
* Is transformation required?

**Deliverables**

* Distribution insights
* Outlier handling strategy (keep / cap / transform)

---

## 🟦 TASK 6: Univariate Analysis – Categorical Features

**What to do**

* Count categories
* Check frequency distribution
* Identify rare categories

**Questions to answer**

* High cardinality?
* Categories with very few samples?

**Deliverables**

* Category frequency summary
* Rare category handling plan

---

## 🟦 TASK 7: Bivariate Analysis – Numerical vs Target

**What to do**

* Compare means/medians by default flag
* Plot boxplots
* Analyze monotonic trends

**Questions to answer**

* Do defaulters behave differently?

**Deliverables**

* Risk-increasing and risk-reducing numerical features

---

## 🟦 TASK 8: Bivariate Analysis – Categorical vs Target

**What to do**

* Compute default rate per category
* Rank categories by risk

**Questions to answer**

* Which categories are high risk?
* Can categories be grouped?

**Deliverables**

* Risk ranking table
* Grouping suggestions

---

## 🟦 TASK 9: Correlation & Multicollinearity Analysis

**What to do**

* Compute correlation matrix
* Identify highly correlated pairs

**Questions to answer**

* Redundant features?
* Logistic regression suitability?

**Deliverables**

* Features to drop or combine

---

## 🟦 TASK 10: Data Leakage Check (CRITICAL)

**What to do**

* Review feature creation time
* Remove post-loan features

**Questions to answer**

* Was this information available at prediction time?

**Deliverables**

* Leakage-free feature list
* Explicit leakage justification

---

## 🟦 TASK 11: Time & Stability Checks (If Applicable)

**What to do**

* Default rate over time
* Feature distribution drift

**Questions to answer**

* Is data stable across time?

**Deliverables**

* Stability observations
* Need for time-based split

---

## 🟦 TASK 12: EDA Summary & Modeling Readiness

**What to do**

* Summarize key findings
* Define next steps

**Deliverables**

* High-risk indicators
* Features to transform/bin
* Features to drop
* Assumptions & risks

---

# ✅ FINAL OUTPUT OF EDA NOTEBOOK

By the end, you should have:

* ✔ Clear understanding of default drivers
* ✔ Clean feature list
* ✔ No leakage
* ✔ Justification for every preprocessing decision

---

If you want next:

* 📓 **Ready-to-use EDA notebook template**
* 🔁 **EDA → Feature Engineering mapping**
* 🎯 **How to explain this EDA in interviews**

Just tell me what you want next 🚀


## Dataset link
-[https://www.kaggle.com/datasets/pradip11/amexpert-codelab-2021?select=train.csv]


# 📊 Dataset Overview

* **Total Records:** ~45,500
* **Problem Type:** Binary Classification
* **Target Variable:** `credit_card_default`

---

# 🧾 1️⃣ Customer Identification

### 1. `customer_id`

* **Type:** Categorical (Unique Identifier)
* **Unique Values:** 45.5k (100%)
* **Missing:** 0%
* **Description:** Unique ID assigned to each customer.
* ⚠️ **Use in Model?** ❌ No (Drop during training)

---

### 2. `name`

* **Type:** Categorical (Nominal)
* **Unique Values:** 4010
* **Missing:** 0%
* **Description:** Customer name.
* ⚠️ **Use in Model?** ❌ No (Irrelevant for prediction)

---

# 👤 2️⃣ Demographic Features

### 3. `age`

* **Type:** Numerical (Continuous)
* **Range:** 23 – 55
* **Mean:** 39
* **Std Dev:** 9.54
* **Missing:** 0%
* **Insight:** Middle-aged majority (Median ≈ 39)

---

### 4. `gender`

* **Type:** Categorical (Binary / Nominal)
* **Distribution:**

  * F → 66%
  * M → 34%
* **Missing:** 0%

---

### 5. `migrant_worker`

* **Type:** Binary (0/1)
* **Mean:** 0.18
* **Interpretation:** 18% are migrant workers
* **Missing:** 0%

---

# 🏠 3️⃣ Household Information

### 6. `owns_car`

* **Type:** Binary
* **True:** 33%
* **False:** 65%
* **Missing:** 1%

---

### 7. `owns_house`

* **Type:** Binary
* **True:** 70%
* **False:** 30%
* **Missing:** 0%

---

### 8. `no_of_children`

* **Type:** Numerical (Discrete)
* **Range:** 0 – 9
* **Mean:** 0.42
* **Missing:** 2%
* **Insight:** Majority have 0 children.

---

### 9. `total_family_members`

* **Type:** Numerical (Discrete)
* **Range:** 1 – 10
* **Mean:** 2.16
* **Missing:** ~0%

---

# 💼 4️⃣ Employment Information

### 10. `occupation_type`

* **Type:** Categorical (Nominal)
* **Unique:** 19 categories
* **Most Common:** Unknown (31%)
* **Missing:** 0%

---

### 11. `no_of_days_employed`

* **Type:** Numerical
* **Range:** 2 – 365k
* **Mean:** 67.6k
* **Missing:** 1%
* ⚠️ Possible outliers (Very high values)

---

# 💰 5️⃣ Financial Information

### 12. `net_yearly_income`

* **Type:** Numerical (Continuous)
* **Range:** 27k – 141M
* **Mean:** 201k
* ⚠️ Strong Outliers Present

---

### 13. `yearly_debt_payments`

* **Type:** Numerical
* **Range:** 2.2k – 328k
* **Mean:** 31.8k
* **Missing:** ~0%

---

### 14. `credit_limit`

* **Type:** Numerical
* **Range:** 4k – 31M
* **Mean:** 43.5k
* ⚠️ Large outlier detected

---

### 15. `credit_limit_used(%)`

* **Type:** Numerical (Percentage)
* **Range:** 0 – 99
* **Mean:** 52.2
* **Std Dev:** 29.4
* **Insight:** Uniformly distributed usage

---

### 16. `credit_score`

* **Type:** Numerical
* **Range:** 500 – 949
* **Mean:** 783
* **Std Dev:** 101
* Strong concentration above 650.

---

# 📉 6️⃣ Credit History Features

### 17. `prev_defaults`

* **Type:** Discrete Numerical
* **Range:** 0 – 2
* **Mean:** 0.06
* Majority have no previous defaults.

---

### 18. `default_in_last_6months`

* **Type:** Binary (0/1)
* **Mean:** 0.05
* 5% default rate recently.

---

# 🎯 7️⃣ Target Variable

### 19. `credit_card_default`

* **Type:** Binary (0/1)
* **Default Rate:** 8%
* **Class Imbalance:** Yes (Imbalanced dataset)

---

# 🧠 Final Structured View (By Data Type)

## 🔢 Numerical Features

* age
* no_of_children
* total_family_members
* no_of_days_employed
* net_yearly_income
* yearly_debt_payments
* credit_limit
* credit_limit_used(%)
* credit_score
* prev_defaults

## 🏷️ Categorical Features

* gender
* occupation_type

## 🔘 Binary Features

* owns_car
* owns_house
* migrant_worker
* default_in_last_6months

## ❌ Drop Features

* customer_id
* name

## 🎯 Target

* credit_card_default


In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [3]:
credit_risk_data = pd.read_csv("/home/hb/creditRiskAnalyzer/data/train.csv")

In [4]:
credit_risk_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 45528 entries, 0 to 45527
Data columns (total 19 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   customer_id              45528 non-null  str    
 1   name                     45528 non-null  str    
 2   age                      45528 non-null  int64  
 3   gender                   45528 non-null  str    
 4   owns_car                 44981 non-null  str    
 5   owns_house               45528 non-null  str    
 6   no_of_children           44754 non-null  float64
 7   net_yearly_income        45528 non-null  float64
 8   no_of_days_employed      45065 non-null  float64
 9   occupation_type          45528 non-null  str    
 10  total_family_members     45445 non-null  float64
 11  migrant_worker           45441 non-null  float64
 12  yearly_debt_payments     45433 non-null  float64
 13  credit_limit             45528 non-null  float64
 14  credit_limit_used(%)     45528 no

### 📊 Data Type Summary

* **Total Rows:** 45,528
* **Total Columns:** 19

**🔤 Categorical (str) – 6 columns**
customer_id, name, gender, owns_car, owns_house, occupation_type

**🔢 Integer (int64) – 5 columns**
age, credit_limit_used(%), prev_defaults, default_in_last_6months, credit_card_default

**🔢 Float (float64) – 8 columns**
no_of_children, net_yearly_income, no_of_days_employed, total_family_members, migrant_worker, yearly_debt_payments, credit_limit, credit_score

### Total missing value

In [6]:
credit_risk_data.isnull().sum()

customer_id                0.000000
name                       0.000000
age                        0.000000
gender                     0.000000
owns_car                   1.201458
owns_house                 0.000000
no_of_children             1.700053
net_yearly_income          0.000000
no_of_days_employed        1.016957
occupation_type            0.000000
total_family_members       0.182305
migrant_worker             0.191091
yearly_debt_payments       0.208663
credit_limit               0.000000
credit_limit_used(%)       0.000000
credit_score               0.017572
prev_defaults              0.000000
default_in_last_6months    0.000000
credit_card_default        0.000000
dtype: float64

In [7]:
credit_risk_data['credit_card_default'].value_counts()

credit_card_default
0    41831
1     3697
Name: count, dtype: int64

In [8]:
credit_risk_data['credit_card_default'].value_counts(normalize=True) * 100

credit_card_default
0    91.879722
1     8.120278
Name: proportion, dtype: float64

- so according to this the data is imbalance from 100% only 8.12% people are defaulter while 91.88% people are not defaulter